planner/cascades/transformation_rules.go

```diff
@@ -2502,3 +2505,133 @@ func (r *PullSelectionUpApply) OnTransform(old *memo.ExprIter) (newExprs []*memo
+	baseRule
+}
+
+// NewRuleMergeAdjacentWindow creates a new Transformation MergeAdjacentWindow.
+// The pattern of this rule is `Window -> Window`.
+func NewRuleMergeAdjacentWindow() Transformation {
+	rule := &MergeAdjacentWindow{}
+	rule.pattern = memo.BuildPattern(
+		memo.OperandWindow,
+		memo.EngineAll,
+		memo.NewPattern(memo.OperandWindow, memo.EngineAll),
+	)
+	return rule
+}
+
+// Match implements Transformation interface.
+func (r *MergeAdjacentWindow) Match(expr *memo.ExprIter) bool {
+	curWinPlan := expr.GetExpr().ExprNode.(*plannercore.LogicalWindow)
+	nextGroupExpr := expr.Children[0].GetExpr()
+	nextWinPlan := nextGroupExpr.ExprNode.(*plannercore.LogicalWindow)
+	nextSchema := nextGroupExpr.Group.Prop.Schema
+	nextGroupChildren := nextGroupExpr.Children
+	ctx := expr.GetExpr().ExprNode.SCtx()
+
+	// Whether Partition parts are the same.
+	if len(curWinPlan.PartitionBy) != len(nextWinPlan.PartitionBy) ||
+		len(curWinPlan.OrderBy) != len(nextWinPlan.OrderBy) ||
+		(curWinPlan.Frame == nil && nextWinPlan.Frame != nil) ||
+		(curWinPlan.Frame != nil && nextWinPlan.Frame == nil) {
+		return false
+	}
+	for i, item := range curWinPlan.PartitionBy {
+		if !item.Col.Equal(ctx, nextWinPlan.PartitionBy[i].Col) ||
+			item.Desc != nextWinPlan.PartitionBy[i].Desc {
+			return false
+		}
+	}
+
+	// Whether OrderBy parts are the same.
+	for i, item := range curWinPlan.OrderBy {
+		if !item.Col.Equal(ctx, nextWinPlan.OrderBy[i].Col) ||
+			item.Desc != nextWinPlan.OrderBy[i].Desc {
+			return false
+		}
+	}
+
+	// Whether Frame parts are the same.
+	if curWinPlan.Frame != nil {
+		if curWinPlan.Frame.Type != nextWinPlan.Frame.Type ||
+			curWinPlan.Frame.Start.Type != nextWinPlan.Frame.Start.Type ||
+			curWinPlan.Frame.Start.UnBounded != nextWinPlan.Frame.Start.UnBounded ||
+			curWinPlan.Frame.Start.Num != nextWinPlan.Frame.Start.Num ||
+			curWinPlan.Frame.End.Type != nextWinPlan.Frame.End.Type ||
+			curWinPlan.Frame.End.UnBounded != nextWinPlan.Frame.End.UnBounded ||
+			curWinPlan.Frame.End.Num != nextWinPlan.Frame.End.Num {
+			return false
+		}
+		for i, expr := range curWinPlan.Frame.Start.CalcFuncs {
+			if !expr.Equal(ctx, nextWinPlan.Frame.Start.CalcFuncs[i]) {
+				return false
+			}
+		}
+		for i, expr := range curWinPlan.Frame.End.CalcFuncs {
+			if !expr.Equal(ctx, nextWinPlan.Frame.End.CalcFuncs[i]) {
+				return false
+			}
+		}
+	}
+
+	// Whether the first window uses the unsettled columns in the next window.
+
+	// `select a, b, sum(bb) over (partition by a) as 'sum_bb' from (select a, b, max(b) over (partition by a) as 'bb' from t) as tt`
+	// The adjacent windows in the above sql statement cannot be merged.
+	// The reason is that the first one uses an unsettled column `bb` from the second one.
+	nextWindowNewCols := make(map[*expression.Column]struct{})
+	for _, nc := range nextSchema.Columns {
+		isExisted := false
+		for _, ngc := range nextGroupChildren {
+			for _, c := range ngc.Prop.Schema.Columns {
```

Comment:
IMO, we don't need to loop twice and reuse the result here. How about following code?
```
	nextWindowCols := make(map[int64]struct{})
	nextWindowChildrenExistedCols := make(map[int64]struct{})
	nextSchema := nextGroupExpr.Group.Prop.Schema
	nextGroupChildren := nextGroupExpr.Children
	for _, ngc := range nextGroupChildren {
		for _, c := range ngc.Prop.Schema.Columns {
			nextWindowChildrenExistedCols[c.UniqueID] = struct{}{}
		}
	}
	for _, nc := range nextSchema.Columns {
		nextWindowCols[nc.UniqueID] = struct{}{}
	}
	for _, funDesc := range curWinPlan.WindowFuncDescs {
		for _, arg := range funDesc.Args {
			cols := expression.ExtractColumns(arg)
			for _, c := range cols {
				_, ok1 := nextWindowCols[c.UniqueID]
				_, ok2 := nextWindowChildrenExistedCols[c.UniqueID]
				if ok1 && !ok2 {
					return false
				}
			}
		}
	}
```

Reply:
I'm not sure that using `UniqueID` here is correct.

Reply:
> I'm not sure that using `UniqueID` here is correct.

Well example. I think using `UniqueID` is correct here because the column compare function `Equal` is implemented by `UniqueID`, too. I will modify that soon.

Reply:
@mmyj I have adjusted the rule area to `TiDBLayerOptimizationBatch` and `Equal` judge codes in [newer commit](https://github.com/pingcap/tidb/pull/24556/commits/071f638ba9a88c5b87f105b7b5693276defaf37a) just now.